File: flutter/flutter_module/lib/src/repo_search_screen.dart
import 'dart:async';

import 'package:flutter/material.dart';

import 'github_client.dart';
import 'host_channel.dart';

const _primary = Color(0xFF111827);
const _secondary = Color(0xFFE5E7EB);
const _subtitle = Color(0xFF6B7280);
const _description = Color(0xFF4B5563);
const _placeholder = Color(0xFF9CA3AF);
const _errorColor = Color(0xFFB91C1C);

/// The screen every implementation embeds: searches GitHub for the keyword the
/// host app handed over and lists the repositories.
class RepoSearchScreen extends StatefulWidget {
  const RepoSearchScreen({
    super.key,
    required this.initialKeyword,
    required this.client,
    required this.host,
    required this.keywordChanges,
  });

  /// Handed over by the host app when it opened the screen.
  final String initialKeyword;
  final GitHubClient client;
  final HostChannel host;

  /// Keywords the host app sends while the screen is open.
  final Stream<String> keywordChanges;

  @override
  State<RepoSearchScreen> createState() => _RepoSearchScreenState();
}

class _RepoSearchScreenState extends State<RepoSearchScreen> {
  late String _keyword = widget.initialKeyword;
  List<Repository> _repositories = const [];
  bool _isLoading = false;
  String? _error;
  StreamSubscription<String>? _keywordSubscription;

  @override
  void initState() {
    super.initState();
    widget.host.markAfterFrame(BenchMarker.embedFirstFrame);
    // The keyword passed in only arrives while the screen is being created,
    // so replacing it on an open screen comes over the channel.
    _keywordSubscription = widget.keywordChanges.listen((next) {
      setState(() {
        _keyword = next;
        _repositories = const [];
        _error = null;
      });
      widget.host.markAfterFrame(BenchMarker.keywordApplied);
    });
  }

  @override
  void dispose() {
    _keywordSubscription?.cancel();
    super.dispose();
  }

  Future<void> _search() async {
    widget.host.mark(BenchMarker.searchTapped);
    final keyword = _keyword;
    setState(() {
      _isLoading = true;
      _error = null;
    });
    try {
      final results = await widget.client.searchRepositories(keyword);
      if (!mounted) return;
      setState(() => _repositories = results);
      if (results.isNotEmpty) {
        widget.host.markAfterFrame(BenchMarker.searchRendered);
      }
      // Hand the results back to the native host app.
      widget.host.notifySearchSucceeded(keyword, results);
    } catch (e) {
      final message = e is GitHubException ? e.message : e.toString();
      if (!mounted) return;
      setState(() {
        _repositories = const [];
        _error = message;
      });
      widget.host.notifySearchFailed(keyword, message);
    } finally {
      if (mounted) setState(() => _isLoading = false);
    }
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.white,
      body: SafeArea(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            Padding(
              padding: const EdgeInsets.fromLTRB(20, 24, 20, 0),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Text(
                    'GitHub Repositories',
                    style: TextStyle(fontSize: 24, fontWeight: FontWeight.w700),
                  ),
                  const SizedBox(height: 4),
                  Text(
                    'keyword: $_keyword',
                    style: const TextStyle(fontSize: 14, color: _subtitle),
                  ),
                ],
              ),
            ),
            Padding(
              padding: const EdgeInsets.symmetric(horizontal: 20, vertical: 16),
              child: Row(
                children: [
                  Expanded(
                    child: _ActionButton(
                      label: _isLoading ? '検索中...' : 'リポジトリを検索',
                      onPressed: _isLoading ? null : _search,
                    ),
                  ),
                  const SizedBox(width: 8),
                  _ActionButton(
                    label: 'ネイティブに戻る',
                    secondary: true,
                    onPressed: widget.host.close,
                  ),
                ],
              ),
            ),
            if (_error != null)
              Padding(
                padding: const EdgeInsets.fromLTRB(20, 0, 20, 8),
                child: Text(
                  _error!,
                  style: const TextStyle(color: _errorColor),
                ),
              ),
            Expanded(child: _buildContent()),
          ],
        ),
      ),
    );
  }

  Widget _buildContent() {
    if (_isLoading && _repositories.isEmpty) {
      return const Align(
        alignment: Alignment.topCenter,
        child: Padding(
          padding: EdgeInsets.only(top: 32),
          child: CircularProgressIndicator(),
        ),
      );
    }
    if (_repositories.isEmpty) {
      return _error == null
          ? const Padding(
              padding: EdgeInsets.only(top: 32),
              child: Text(
                'ボタンを押すと検索結果が表示されます。',
                textAlign: TextAlign.center,
                style: TextStyle(color: _placeholder),
              ),
            )
          : const SizedBox.shrink();
    }
    return ListView.builder(
      padding: const EdgeInsets.fromLTRB(20, 0, 20, 32),
      itemCount: _repositories.length,
      itemBuilder: (context, index) =>
          _RepositoryRow(repository: _repositories[index]),
    );
  }
}

class _ActionButton extends StatelessWidget {
  const _ActionButton({
    required this.label,
    required this.onPressed,
    this.secondary = false,
  });

  final String label;
  final VoidCallback? onPressed;
  final bool secondary;

  @override
  Widget build(BuildContext context) {
    return FilledButton(
      onPressed: onPressed,
      style: FilledButton.styleFrom(
        backgroundColor: secondary ? _secondary : _primary,
        foregroundColor: secondary ? _primary : Colors.white,
        disabledBackgroundColor: _primary.withValues(alpha: 0.5),
        disabledForegroundColor: Colors.white,
        padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
        shape: RoundedRectangleBorder(borderRadius: BorderRadius.circular(10)),
      ),
      child: Text(
        label,
        style: const TextStyle(fontSize: 15, fontWeight: FontWeight.w600),
      ),
    );
  }
}

class _RepositoryRow extends StatelessWidget {
  const _RepositoryRow({required this.repository});

  final Repository repository;

  @override
  Widget build(BuildContext context) {
    final language = repository.language;
    return DecoratedBox(
      decoration: const BoxDecoration(
        border: Border(top: BorderSide(color: _secondary, width: 0.5)),
      ),
      child: Padding(
        padding: const EdgeInsets.symmetric(vertical: 12),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              repository.fullName,
              style: const TextStyle(fontSize: 16, fontWeight: FontWeight.w600),
            ),
            if (repository.description != null) ...[
              const SizedBox(height: 4),
              Text(
                repository.description!,
                maxLines: 2,
                overflow: TextOverflow.ellipsis,
                style: const TextStyle(fontSize: 13, color: _description),
              ),
            ],
            const SizedBox(height: 6),
            Text(
              '★ ${groupThousands(repository.stars)}'
              '${language != null ? ' · $language' : ''}',
              style: const TextStyle(fontSize: 12, color: _subtitle),
            ),
          ],
        ),
      ),
    );
  }
}

/// 51842 → "51,842", as the other implementations format star counts.
String groupThousands(int value) => value.toString().replaceAllMapped(
  RegExp(r'\B(?=(\d{3})+(?!\d))'),
  (_) => ',',
);

File: kmp/shared/src/commonMain/kotlin/com/example/benchmark/kmp/reposearchkit/RepoSearchScreen.kt
package com.example.benchmark.kmp.reposearchkit

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.HorizontalDivider
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.withFrameNanos
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlinx.coroutines.launch

/**
 * The screen every implementation embeds: searches GitHub for the keyword the
 * host app handed over and lists the repositories. One Compose Multiplatform
 * source for both platforms.
 */
@Composable
fun RepoSearchScreen(
  keyword: String,
  onClose: () -> Unit,
  modifier: Modifier = Modifier,
  apiBaseUrl: String = GitHubClient.DEFAULT_BASE_URL,
) {
  val model = remember { RepoSearchModel(keyword, GitHubClient(apiBaseUrl)) }
  val scope = rememberCoroutineScope()
  DisposableEffect(model) { onDispose { model.dispose() } }

  // withFrameNanos resumes on the frame after the one being built: the first
  // one draws the change, the second means it is on screen. Every
  // implementation waits for the same two frame boundaries (see AGENTS.md).
  LaunchedEffect(Unit) {
    withFrameNanos {}
    withFrameNanos {}
    BenchMarker.mark("embedFirstFrame")
  }
  LaunchedEffect(model.repositories) {
    if (model.repositories.isNotEmpty()) {
      withFrameNanos {}
      withFrameNanos {}
      BenchMarker.mark("searchRendered")
    }
  }
  val initialKeyword = remember { keyword }
  LaunchedEffect(model.keyword) {
    if (model.keyword != initialKeyword) {
      withFrameNanos {}
      withFrameNanos {}
      BenchMarker.mark("keywordApplied")
    }
  }

  Column(modifier = modifier.fillMaxSize().background(Color.White)) {
    Column(modifier = Modifier.padding(start = 20.dp, top = 24.dp, end = 20.dp)) {
      Text("GitHub Repositories", fontSize = 24.sp, fontWeight = FontWeight.Bold)
      Text(
        "keyword: ${model.keyword}",
        color = Palette.subtitle,
        fontSize = 14.sp,
        modifier = Modifier.padding(top = 4.dp),
      )
    }

    Row(
      horizontalArrangement = Arrangement.spacedBy(8.dp),
      modifier = Modifier.padding(horizontal = 20.dp, vertical = 16.dp),
    ) {
      ActionButton(
        label = if (model.isLoading) "検索中..." else "リポジトリを検索",
        enabled = !model.isLoading,
        onClick = {
          BenchMarker.mark("searchTapped")
          scope.launch { model.search() }
        },
        modifier = Modifier.weight(1f),
      )
      ActionButton(label = "ネイティブに戻る", secondary = true, onClick = onClose)
    }

    model.errorMessage?.let { message ->
      Text(
        message,
        color = Palette.error,
        modifier = Modifier.padding(start = 20.dp, end = 20.dp, bottom = 8.dp),
      )
    }

    when {
      model.isLoading && model.repositories.isEmpty() ->
        CircularProgressIndicator(
          modifier = Modifier.padding(top = 32.dp).align(Alignment.CenterHorizontally)
        )
      model.repositories.isEmpty() ->
        if (model.errorMessage == null) {
          Text(
            "ボタンを押すと検索結果が表示されます。",
            color = Palette.placeholder,
            textAlign = TextAlign.Center,
            modifier = Modifier.fillMaxWidth().padding(top = 32.dp),
          )
        }
      else ->
        LazyColumn(contentPadding = PaddingValues(start = 20.dp, end = 20.dp, bottom = 32.dp)) {
          items(model.repositories, key = { it.id }) { repository ->
            RepositoryRow(repository)
          }
        }
    }
  }
}

private object Palette {
  val primary = Color(0xFF111827)
  val secondary = Color(0xFFE5E7EB)
  val subtitle = Color(0xFF6B7280)
  val description = Color(0xFF4B5563)
  val placeholder = Color(0xFF9CA3AF)
  val error = Color(0xFFB91C1C)
}

@Composable
private fun ActionButton(
  label: String,
  onClick: () -> Unit,
  modifier: Modifier = Modifier,
  secondary: Boolean = false,
  enabled: Boolean = true,
) {
  Button(
    onClick = onClick,
    enabled = enabled,
    shape = RoundedCornerShape(10.dp),
    colors =
      ButtonDefaults.buttonColors(
        containerColor = if (secondary) Palette.secondary else Palette.primary,
        contentColor = if (secondary) Palette.primary else Color.White,
        disabledContainerColor = Palette.primary.copy(alpha = 0.5f),
        disabledContentColor = Color.White,
      ),
    contentPadding = PaddingValues(horizontal = 16.dp, vertical = 12.dp),
    modifier = modifier,
  ) {
    Text(label, fontSize = 15.sp, fontWeight = FontWeight.SemiBold)
  }
}

@Composable
private fun RepositoryRow(repository: Repository) {
  Column {
    HorizontalDivider(thickness = 0.5.dp, color = Palette.secondary)
    Column(modifier = Modifier.padding(vertical = 12.dp)) {
      Text(repository.fullName, fontSize = 16.sp, fontWeight = FontWeight.SemiBold)
      repository.description?.let { description ->
        Text(
          description,
          color = Palette.description,
          fontSize = 13.sp,
          maxLines = 2,
          overflow = TextOverflow.Ellipsis,
          modifier = Modifier.padding(top = 4.dp),
        )
      }
      Text(
        buildString {
          append("★ ${groupThousands(repository.stars)}")
          repository.language?.let { append(" · $it") }
        },
        color = Palette.subtitle,
        fontSize = 12.sp,
        modifier = Modifier.padding(top = 6.dp),
      )
    }
  }
}

/** 51842 → "51,842". Common code has no NumberFormat. */
internal fun groupThousands(value: Int): String =
  value.toString().reversed().chunked(3).joinToString(",").reversed()
